### ISSUE TITLE
shadings are no longer defined globally

### ISSUE THREAD

**Issue by u-fischer** [pgf-tikz/pgf#650]
In `Package: pgf 2019/02/02 v3.1.1 (3.1.1)` the example below where a shading is defined in a group gave

~~~
> \@pgfshadingbeamer@barshade!=macro:
->\unhbox \voidb@x \pdfrefxform 4.
<recently read> \@pgfshadingbeamer@barshade! 
~~~

with  pgf 2019/04/04 v3.1.2 (3.1.2) the command is no longer defined:

~~~
\@pgfshadingbeamer@barshade!=\relax.
~~~

Removing the additional group resolves the problem, but as some beamer themes (e.g. smoothbar) define shadings in a group they fail now. 

~~~
\documentclass{article}
\usepackage{pgf}

{ %<--- group
\pgfdeclareverticalshading{beamer@barshade}{\the\paperwidth}{%
         color(0ex)=(red);%
         color(1ex)=(green);%
         color(7ex)=(blue)%
       }
}       

\expandafter\show \csname @pgfshadingbeamer@barshade!\endcsname

\begin{document}
bb
\end{document}
~~~

**Comment by hmenke** [comment]
Offending commit https://github.com/pgf-tikz/pgf/commit/55cc7bca878f64812b1a4ec91709f3f4d30d63c8

**Comment by domenjoud** [comment]
No, no, no!!!! Shadings do not have to be defined globally! I worked hard bringing some coherence in the definition of shadings. Finally defining them globally just destroys my efforts and simply makes my patches useless. If some badly designed beamer classes assume that a locally defined shading is actually defined globally, it is simply a TeX nonsense. Please correct the beamer class. Do not break again shading definitions in pgf.

**Comment by hmenke** [comment]
@domenjoud Can you please open a bug report at https://github.com/josephwright/beamer/issues?
Once this is fixed in `beamer` I will revert https://github.com/pgf-tikz/pgf/commit/74d48c897469333c25b96031711bf36f22a9e4f4.

**Comment by josephwright** [comment]
@domenjoud Could you explain why you've wanted to make shadings local? Aren't they defined in dictionary for PDF on a per-page or per-document basis? (Similar for PostScript). I can adjust `beamer`, but before I do I'd like a little background.

**Comment by domenjoud** [comment]
@josephwright I do not insist on shadings being defined locally. The problem is that shading declaration in pgf was not coherent in the sense that a shading with parameters was defined locally while its instances, and also parameterless shadings, were defined globally. It made impossible to redefine a shading because it produced almost unpredictible results (see ticket #510). Besides, when something is defined inside a TeX group, one might expect it to remain local. pdf and postscript shadings are indeed defined globally but it does not mean that the TeX macro invoking them has to be global. If a pgf shading is redefined, a new pdf or ps shading will be (globally) defined and the TeX macro will invoke this new definition until the end of the current group.

**Comment by josephwright** [comment]
@domenjoud Ah, OK: I know there are various scope issues I've spotted in `pgf` in the past. I'll look to make the `beamer` cases globally-defined (I think they should be there). Thanks.

**Comment by hmenke** [comment]
For completeness: https://github.com/josephwright/beamer/issues/536

**Comment by hmenke** [comment]
@domenjoud There was another issue with local shadings, see https://github.com/pgf-tikz/pgf/commit/bbe9ae5412b08192ef887af136ba743dd33b25da
